src/dbnode/persist/fs/retriever.go

```diff
@@ -299,12 +306,19 @@ func (r *blockRetriever) fetchBatch(
 
 	// Sort the requests by offset into the file before seeking
 	// to ensure all seeks are in ascending order
+	var entrySize int
 	for _, req := range reqs {
+		if err := r.opts.QueryLimits().BytesReadLimit().Inc(entrySize); err != nil {
```

Comment:
If a limit is tripped, `entrySize` will always be the last tripping value, so all further entries will be logged as being  that size; either we capture the error condition and stop updating these, or we need to do the check after seeking index entry to determine the actual size.

Also there's currently a racey condition, where the limit resets during this fetch. That can lead to conditions where a few entries in the middle of the batch will fail limit checks, but the first and last ones will succeed.

I think the error should be captured, and once it's not nil it shouldn't be updated again, forcing all subsequent entries to fail.

Reply:
Yeah good call. Sharing the entrySize was my attempt at ensuring the whole batch fails once the first failure happens. But you're right that then inflates the total exceeded metrics so we should share the error.